File: services/ai/src/docmind_ai/contracts/schema.py
from typing import Annotated, Literal, Self
from uuid import UUID

from pydantic import Field, JsonValue, model_validator

from docmind_ai.contracts.common import StrictModel

SchemaValueType = Literal[
    "string", "number", "integer", "boolean", "date", "datetime", "object", "array"
]
FieldSensitivity = Literal["none", "low", "medium", "high"]


class NoSchemaFieldDefault(StrictModel):
    kind: Literal["none"]


class LiteralSchemaFieldDefault(StrictModel):
    kind: Literal["literal"]
    value: JsonValue


SchemaFieldDefault = Annotated[
    NoSchemaFieldDefault | LiteralSchemaFieldDefault, Field(discriminator="kind")
]


class SchemaFieldConstraints(StrictModel):
    format: str | None = None
    pattern: str | None = Field(default=None, max_length=2000)
    enum_values: list[JsonValue] = Field(default_factory=list, max_length=200)
    min_length: int | None = Field(default=None, ge=0)
    max_length: int | None = Field(default=None, ge=0)
    minimum: float | None = None
    maximum: float | None = None

    @model_validator(mode="after")
    def validate_bounds(self) -> Self:
        if (
            self.min_length is not None
            and self.max_length is not None
            and self.min_length > self.max_length
        ):
            raise ValueError("min_length cannot exceed max_length")
        if self.minimum is not None and self.maximum is not None and self.minimum > self.maximum:
            raise ValueError("minimum cannot exceed maximum")
        return self


class SchemaFieldDefinition(StrictModel):
    id: UUID
    key: str = Field(pattern=r"^[A-Za-z_][A-Za-z0-9_]*$", max_length=64)
    json_path: str = Field(
        pattern=r"^\$(?:\.[A-Za-z_][A-Za-z0-9_]*)+$", min_length=3, max_length=500
    )
    description: str = Field(max_length=2000)
    value_type: SchemaValueType
    array_item_type: SchemaValueType | None
    required: bool
    nullable: bool
    default: SchemaFieldDefault
    sensitivity: FieldSensitivity
    constraints: SchemaFieldConstraints
    examples: list[JsonValue] = Field(default_factory=list, max_length=20)
    extraction_hint: str | None = Field(default=None, max_length=2000)
    position: int = Field(ge=0)

File: apps/docmind-web/src/contracts/schema/schema.ts
import type {
  AuditMetadata,
  IsoDateTime,
  JsonObject,
  JsonValue,
  SchemaFieldId,
  SchemaId,
  SchemaTemplateId,
  SchemaVersionId,
  UserId,
  WorkspaceId,
} from '../common/index.js';

export const SCHEMA_VALUE_TYPES = [
  'string',
  'number',
  'integer',
  'boolean',
  'date',
  'datetime',
  'object',
  'array',
] as const;
export type SchemaValueType = (typeof SCHEMA_VALUE_TYPES)[number];

export const FIELD_SENSITIVITY_LEVELS = ['none', 'low', 'medium', 'high'] as const;
export type FieldSensitivityLevel = (typeof FIELD_SENSITIVITY_LEVELS)[number];

export const SCHEMA_VERSION_STATUSES = ['draft', 'published', 'superseded'] as const;
export type SchemaVersionStatus = (typeof SCHEMA_VERSION_STATUSES)[number];

export type SchemaFieldDefault = { kind: 'none' } | { kind: 'literal'; value: JsonValue };

export interface SchemaFieldConstraints {
  format: string | null;
  pattern: string | null;
  enum_values: JsonValue[];
  min_length: number | null;
  max_length: number | null;
  minimum: number | null;
  maximum: number | null;
}

export interface SchemaFieldDisplayOptions {
  /** Presentation-only masking instruction; never changes the stored extracted value. */
  mask: 'none' | 'partial' | 'full';
  /** Role keys granted access to the unmasked field value. */
  view_role_keys: string[];
}

export interface SchemaFieldDefinition {
  id: SchemaFieldId;
  key: string;
  json_path: string;
  description: string;
  value_type: SchemaValueType;
  array_item_type: SchemaValueType | null;
  required: boolean;
  nullable: boolean;
  default: SchemaFieldDefault;
  sensitivity: FieldSensitivityLevel;
  constraints: SchemaFieldConstraints;
  examples: JsonValue[];
  extraction_hint: string | null;
  display: SchemaFieldDisplayOptions;
  /** Non-executable tenant metadata for future UI and workflow extensions. */
  metadata: JsonObject;
  position: number;
}

/** Client-supplied field definition. Entity IDs are allocated by the service. */
export interface SchemaFieldInput {
  key: string;
  json_path: string;
  description: string;
  value_type: SchemaValueType;
  array_item_type: SchemaValueType | null;
  required: boolean;
  nullable: boolean;
  default: SchemaFieldDefault;
  sensitivity: FieldSensitivityLevel;
  constraints: SchemaFieldConstraints;
  examples: JsonValue[];
  extraction_hint: string | null;
  display: SchemaFieldDisplayOptions;
  metadata: JsonObject;
  position: number;
}

export interface ExtractionSchema extends AuditMetadata {
  id: SchemaId;
  workspace_id: WorkspaceId;
  name: string;
  description: string;
  current_version_id: SchemaVersionId | null;
}

export interface SchemaVersion {
  id: SchemaVersionId;
  schema_id: SchemaId;
  workspace_id: WorkspaceId;
  version_number: number;
  status: SchemaVersionStatus;
  fields: SchemaFieldDefinition[];
  json_schema: JsonObject;
  change_summary: string;
  created_at: IsoDateTime;
  created_by: UserId;
  published_at: IsoDateTime | null;
}

/** Reusable pointer to an immutable, published schema version snapshot. */
export interface SchemaTemplate extends AuditMetadata {
  id: SchemaTemplateId;
  workspace_id: WorkspaceId;
  name: string;
  description: string;
  current_schema_version_id: SchemaVersionId;
}

export interface CreateSchemaRequest {
  name: string;
  description: string;
  fields: SchemaFieldInput[];
}

export interface CreateSchemaVersionRequest {
  fields: SchemaFieldInput[];
  change_summary: string;
}

export interface CreateSchemaTemplateRequest {
  name: string;
  description: string;
  schema_version_id: SchemaVersionId;
}
